src/testing/storage.zig

```diff
@@ -461,8 +573,57 @@ pub const Storage = struct {
     }
 
     fn write_sectors_finish(storage: *Storage, write: *Storage.Write) void {
+        assert(storage.overlays.total() >= 2);
+
         hash_log.emit_autohash(.{ write.buffer, write.zone, write.offset }, .DeepRecursive);
 
+        // Clean up old misdirects if they are overwritten.
+        var overlays_iterator = storage.overlays.iterate();
+        while (overlays_iterator.next()) |overlay| {
+            if (overlay.zone == write.zone and
+                overlay.offset == write.offset)
+            {
+                storage.overlays.release(overlay);
+            }
+        }
+
+        // Apply a new misdirect.
+        const misdirect = storage.overlays.available() >= 2 and
+            storage.pick_faulty_sector(write.zone, write.offset, write.buffer.len) != null and
+            storage.x_in_100(storage.options.write_misdirect_probability);
+        const misdirect_offset = if (misdirect) storage.pick_faulty_chunk_offset(write) else null;
+        if (misdirect_offset) |mistaken_offset| {
+            assert(mistaken_offset != write.offset);
+
+            const overlay_mistaken = storage.overlays.acquire().?;
+            const overlay_intended = storage.overlays.acquire().?;
+
+            const overlay_mistaken_index = storage.overlays.index(overlay_mistaken);
+            const overlay_intended_index = storage.overlays.index(overlay_intended);
+
+            log.debug("{}: write_sectors_finish: misdirect zone={s} offset={}->{} size={}", .{
+                storage.options.replica_index.?,
+                @tagName(write.zone),
+                write.offset,
+                mistaken_offset,
+                write.buffer.len,
+            });
+
+            const overlay_size: u32 = @intCast(write.buffer.len);
+            overlay_mistaken.* =
+                .{ .zone = write.zone, .offset = mistaken_offset, .size = overlay_size };
+            overlay_intended.* =
+                .{ .zone = write.zone, .offset = write.offset, .size = overlay_size };
+
+            const overlay_mistaken_buffer = storage.overlay_buffers[overlay_mistaken_index];
+            const overlay_intended_buffer = storage.overlay_buffers[overlay_intended_index];
+            const target_intended_buffer =
+                storage.memory[write.zone.offset(write.offset)..][0..write.buffer.len];
+
+            stdx.copy_disjoint(.inexact, u8, overlay_mistaken_buffer, write.buffer);
+            stdx.copy_disjoint(.inexact, u8, overlay_intended_buffer, target_intended_buffer);
+        }
+
         const offset_in_storage = write.zone.offset(write.offset);
         stdx.copy_disjoint(
```

Comment:
Maybe move this copy to happen right before we return? So that we first handle two-faults, and setup pristine memory last.

Reply:
Incidentally, `pristine` might be a better name for `memory` perhaps?

Reply:
I slightly prefer `memory`, mostly because it pairs well with `memory_written`.